### ISSUE TITLE
Embedding layer about cnn_text_classification

### ISSUE THREAD

**Issue by lan2720** [dmlc/mxnet#2467]
@qcl6355 Hi，thank you for your code implementing cnn for text classification. It guides me to do some work with mxnet succesfully!
I'm reading your text_cnn.py. In this code, you've commented two lines about Embedding Layer. 

```    
# embedding layer    
# embed_layer = mx.sym.Embedding(data=input_x, input_dim=vocab_size, output_dim=num_embed, name='vocab_embed')
# embed_layer = mx.sym.Reshape(data=embed_layer, target_shape=(1, 1, sentence_size, num_embed))
```
If I want to try cnn non-static(modify word embedding during training process), I should use embedding layer like these two lines, right?  
But there are two question in my mind,

 1. mx.sym.Embedding() has a param named "weight". Doc says it's embedding weight matrix. For example, if I have 50000 words in my vocabulary, every word will embedded in 100-d vector, the weight matrix should be like (50000, 100) matrix. In the cnn-rand model, the weight matrix will be generated by `np.random.uniform(-a, a, (50000, 100))`, am I right? But why did you build an Embedding layer without **weight**? It initializes a random weight matrix implictly for user?

 2. In mx.sym.Reshape(), the target_shape is (1,1,sentence_size, num_embed). I'm comfused why target_shape[0] = 1. Isn't it (**-1**, 1, sentence_size, num_embed). Because I think the target shape should be (total_sentence, 1, sentence_maxlen, num_embed).

If I got something wrong, pls correct me! Look forward to your answers: )

**Comment by qcl6355** [comment]
1. The embedding layer will update its weight parameters in the training loop. You can think this process as learning word vectors automatically.
2. As you said, if using batched learning, the target shape should be (batch_size, 1, sentence_maxlen, num_embed).

I am sorry for leaving these ambiguous lines in the original code. Recently, I have updated the code about auto learning embeddings in my own forked repo. Soon, I will update this example into dmlc/mxnet.

Thanks for your trial.

**Comment by lan2720** [comment]
@qcl6355 
Thanks! Pretty good work!

**Comment by zihaolucky** [comment]
Hi @lan2720 I used to explore the network structure, the size issue here using `
```python
input_shapes['data'] = (batch_size, sentence_size)  # change here to see if things are going well.
conv_input.infer_shape(**input_shapes)
```

```python
  1 import mxnet as mx
  2 import numpy as np
  3
  4 # imperative and symbolic programming.
  5 # cnn text classification
  6
  7 # some setups
  8 vocab_size = 10000  # assume our corpus has 10,000 unique words
  9 embedding_size = 128  # word vector size
 10 sentence_size = 10  # 10 words in this sentence
 11 batch_size = 32
 12
 13 x = mx.symbol.Variable('data')
 14 y = mx.symbol.Variable('softmax_label')
 15
 16 # embedding layer. if not with_embedding.
 17 embed_layer = mx.symbol.Embedding(data = x,
 18                                   input_dim = vocab_size,
 19                                   output_dim = embedding_size,
 20                                   name = 'word_embedding')
 21 conv_input = mx.symbol.Reshape(data = embed_layer,
 22                                shape = (batch_size, 1, sentence_size, embedding_size))
```

**Comment by zihaolucky** [comment]
Hi @qcl6355 

The example you contributed help me learned a lot about the framework and api. I'm now wondering whether there's any possible to use more high-level api, which just simply define layers and avoid explicitly figuring out network's size.

Another question is, I notice the code doesn't support multichannel setting as the original paper did, in which it only updates one of the two channels.

**Comment by qcl6355** [comment]
Of course you can.
For example, you could use more high-level api to write an data iterator to do some padding. Maybe you can use module api to leverage multi-gpu.

The second question,  notice that the _target_shape_ parameter in the reshape operator, which is (batch_size, channel, sentence_size, num_embed), you can prepare your data in this format, i.e., put the trained wordvecs data in the second channel.

**Comment by zihaolucky** [comment]
@qcl6355 Yep. I've already used `DataIter` for data loading. 

Here's my plan on the second, I prepare the first, static, channel with an embedding layer initialized with word2vec. The embedding layer is used to generate `(batch_size, 1, sentence_size, num_embed)` input. The second is identical to the first, then `concat` them into `(batch_size, 2, sentence_size, num_embed)` as input before `conv`, but only update on the second non-static layer using some fine-tuning tricks as mentioned here https://github.com/dmlc/mxnet/issues/2242
